components/tabs/src/storage.rs

```diff
@@ -146,29 +149,36 @@ impl TabsStorage {
         self.local_tabs.borrow_mut().replace(local_state);
     }
 
+    // We try our best to fit as many tabs in a payload as possible, this includes
+    // limiting the url history entries, title character count and finally drop enough tabs
+    // until we have small enough payload that the server will accept
     pub fn prepare_local_tabs_for_upload(&self) -> Option<Vec<RemoteTab>> {
         if let Some(local_tabs) = self.local_tabs.borrow().as_ref() {
-            return Some(
-                local_tabs
-                    .iter()
-                    .cloned()
-                    .filter_map(|mut tab| {
-                        if tab.url_history.is_empty() || !is_url_syncable(&tab.url_history[0]) {
-                            return None;
+            let mut sanitized_tabs: Vec<RemoteTab> = local_tabs
+                .iter()
+                .cloned()
+                .filter_map(|mut tab| {
+                    if tab.url_history.is_empty() || !is_url_syncable(&tab.url_history[0]) {
+                        return None;
+                    }
+                    let mut sanitized_history = Vec::with_capacity(TAB_ENTRIES_LIMIT);
+                    for url in tab.url_history {
+                        if sanitized_history.len() == TAB_ENTRIES_LIMIT {
+                            break;
                         }
-                        let mut sanitized_history = Vec::with_capacity(TAB_ENTRIES_LIMIT);
-                        for url in tab.url_history {
-                            if sanitized_history.len() == TAB_ENTRIES_LIMIT {
-                                break;
-                            }
-                            if is_url_syncable(&url) {
-                                sanitized_history.push(url);
-                            }
+                        if is_url_syncable(&url) {
+                            sanitized_history.push(url);
                         }
-                        tab.url_history = sanitized_history;
-                        Some(tab)
-                    })
-                    .collect(),
+                    }
+
+                    tab.url_history = sanitized_history;
+                    tab.title.truncate(MAX_TITLE_CHAR_LENGTH);
```

Comment:
I do think we should add an ellipses - the client can't know whether we trimmed it or not so it can't add one. I believe it's likely to be seen in the tooltip for tabs at 512 chars on desktop devices. Another option would be to add a new bool to indicate if it was truncated so the client can add it, but that doesn't seem necessary.